src/main/java/org/folio/service/pieces/flows/update/PieceUpdateFlowManager.java

```diff
@@ -92,10 +111,49 @@ public Future<Void> updatePiece(Piece pieceToUpdate, boolean createItem, boolean
   }
 
   protected Future<Void> updatePoLine(PieceUpdateHolder holder, RequestContext requestContext) {
-    if (!Boolean.TRUE.equals(holder.getOriginPoLine().getIsPackage()) && !Boolean.TRUE.equals(holder.getOriginPoLine().getCheckinItems())) {
-      return updatePoLineService.updatePoLine(holder, requestContext);
-    }
-    return Future.succeededFuture();
+    CompositePoLine originPoLine = holder.getOriginPoLine();
+
+    return getPiecesByPoLine(originPoLine.getId(), requestContext)
+      .map(pieces -> {
+        CompositePurchaseOrder order = holder.getOriginPurchaseOrder();
+        if (order.getOrderType() != OrderType.ONE_TIME || order.getWorkflowStatus() != WorkflowStatus.OPEN) {
+          return Future.succeededFuture();
+        }
+        List<Piece> piecesToUpdate = List.of(holder.getPieceToUpdate());
+        CompositePoLine poLineToSave = holder.getPoLineToSave();
+        poLineToSave.setReceiptStatus(calculatePoLineReceiptStatus(originPoLine, pieces, piecesToUpdate));
+        return purchaseOrderLineService.saveOrderLine(poLineToSave, requestContext);
+      }).compose(t -> {
+        if (!Boolean.TRUE.equals(originPoLine.getIsPackage()) &&
+          !Boolean.TRUE.equals(originPoLine.getCheckinItems())) {
+          return updatePoLineService.updatePoLine(holder, requestContext);
+        }
+        return Future.succeededFuture();
+      });
+  }
+
+  CompositePoLine.ReceiptStatus calculatePoLineReceiptStatus(CompositePoLine poLine, List<Piece> fromStorage, List<Piece> toUpdate) {
+
+    // 1. collect all piece statuses
+    Map<String, Piece.ReceivingStatus> map = new HashMap<>();
+    fromStorage.forEach(piece -> map.put(piece.getId(), piece.getReceivingStatus()));
+    toUpdate.forEach(piece -> map.put(piece.getId(), piece.getReceivingStatus()));
+
+    // 2. count received and expected statuses
+    long receivedQuantity = map.values().stream().filter(RECEIVED_STATUSES::contains).count();
+    long expectedQuantity = map.values().stream().filter(EXPECTED_STATUSES::contains).count();
+
+    logger.info("calculatePoLineReceiptStatus:: POL: {}, received: {}, expected: {}",
+      poLine.getId(), receivedQuantity, expectedQuantity);
+
+    return expectedQuantity == 0 ? FULLY_RECEIVED :
+      receivedQuantity > 0 ? PARTIALLY_RECEIVED : AWAITING_RECEIPT;
+  }
+
+  private Future<List<Piece>> getPiecesByPoLine(String poLineId, RequestContext requestContext) {
```

Comment:
we already have similar method in PieceStorageService: https://github.com/folio-org/mod-orders/blob/master/src/main/java/org/folio/service/pieces/PieceStorageService.java#L47C66-L47C73
Please avoid duplication

Reply:
removed